dapr_agents/observability/context_propagation.py

```diff
@@ -0,0 +1,205 @@
+    OPENTELEMETRY_AVAILABLE = False
+
+logger = logging.getLogger(__name__)
+
+# Global W3C Trace Context propagator instance for consistent context handling
+if OPENTELEMETRY_AVAILABLE:
+    _propagator = TraceContextTextMapPropagator()
+else:
+    _propagator = None
+
+
+def extract_otel_context() -> Dict[str, Any]:
+    """
+    Extract current OpenTelemetry context using W3C Trace Context format.
+
+    Converts the current OpenTelemetry span context into W3C Trace Context format
+    (traceparent/tracestate) that can be properly serialized and transmitted across
+    Dapr Workflow boundaries. This ensures distributed trace continuity.
+
+    W3C traceparent format: "version-trace_id-parent_id-trace_flags"
+    - version: Always "00" (current W3C spec version)
+    - trace_id: 32-character hex string (128-bit trace identifier)
+    - parent_id: 16-character hex string (64-bit span identifier)
+    - trace_flags: 2-character hex string ("01" if sampled, "00" if not)
+
+    Returns:
+        Dict[str, Any]: Serializable context data containing:
+                       - traceparent: W3C trace context header
+                       - tracestate: Vendor-specific trace state (optional)
+                       - trace_id, span_id, trace_flags: Individual components for debugging
+
+        If OpenTelemetry is not available, returns empty dict.
+        Install with: pip install dapr-agents[observability]
+    """
+    if not OPENTELEMETRY_AVAILABLE:
+        logger.debug("OpenTelemetry not available - cannot extract context")
+        return {}
+
+    if not _propagator or not trace:
+        logger.debug(
+            "OpenTelemetry components not initialized - cannot extract context"
+        )
+        return {}
+
+    carrier: Dict[str, Any] = {}
+    # Use W3C TraceContext propagator to inject standard headers
+    _propagator.inject(carrier)
+
+    span = trace.get_current_span()
+    ctx = span.get_span_context()
+
+    # Extract trace components in W3C format (always extract for consistency)
+    trace_id = format(ctx.trace_id, "032x")  # 32-char hex (128-bit)
+    span_id = format(ctx.span_id, "016x")  # 16-char hex (64-bit)
+    flags = "01" if ctx.trace_flags.sampled else "00"  # W3C sampling flag
+
+    # Ensure traceparent exists - create manually if propagator didn't inject
+    # This handles edge cases where OpenTelemetry context might be incomplete
+    if "traceparent" not in carrier and span and span.is_recording():
+        carrier["traceparent"] = f"00-{trace_id}-{span_id}-{flags}"
+
+    # Ensure tracestate exists in carrier (W3C spec allows empty tracestate)
+    if "tracestate" not in carrier:
+        carrier["tracestate"] = ""
+
+    # Add individual components for debugging and verification
+    carrier["trace_id"] = trace_id
+    carrier["span_id"] = span_id
+    carrier["trace_flags"] = flags
+
+    logger.debug(
+        f"🔗 Extracted W3C context: traceparent={carrier.get('traceparent', '')}"
+    )
```

Comment:
we can probably rm this

Reply:
actually for debugging purposes, I think we should keep it.